Orm/Xtensive.Orm/Orm/Providers/SqlCompiler.Index.cs

```diff
@@ -7,16 +7,38 @@
 using System;
 using System.Collections.Generic;
 using System.Linq;
+using System.Linq.Expressions;
+using Xtensive.Core;
+using Xtensive.Linq;
+using Xtensive.Orm.Internals;
 using Xtensive.Orm.Model;
 using Xtensive.Orm.Rse.Providers;
+using Xtensive.Reflection;
 using Xtensive.Sql;
 using Xtensive.Sql.Dml;
 using IndexInfo = Xtensive.Orm.Model.IndexInfo;
 
 namespace Xtensive.Orm.Providers
 {
-  partial class SqlCompiler 
+  public partial class SqlCompiler
   {
+    protected readonly struct QueryAndBindings
+    {
+      public SqlSelect Query { get; }
+      public List<QueryParameterBinding> Bindings { get; }
```

Comment:
Why this property is of the `List<>` type? Do we add something to this list after the `QueryAndBindings` struct instance construction? If no, this property type could be changed to e.g `IReadOnlyCollection` or `IReadOnlyList`.

Reply:
List is used because there are some cases when we need to merge bindings from several queries. For instance, ```SqlCompiler.BuildUnionQuery```, It builds several base queries and each of them may have bindings, I chose to use first query bindings as result bindings and add the rest of bindings to it instead of creating new collection every single time I need to merge bindings.